golem-rust/src/transaction/mod.rs

```diff
@@ -0,0 +1,559 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod compfn;
+
+use std::fmt::{Debug, Display, Formatter};
+use std::rc::Rc;
+
+use crate::bindings::golem::api::host::{get_oplog_index, set_oplog_index, OplogIndex};
+use crate::mark_atomic_operation;
+
+pub use compfn::*;
+
+/// Represents an atomic operation of the transaction which has a rollback action.
+///
+/// Implement this trait and use it within a `transaction` block.
+/// Operations can also be constructed from closures using `operation`.
+pub trait Operation: Clone {
+    type In: Clone;
+    type Out;
+    type Err;
+
+    /// Executes the operation which may fail with a domain error
+    fn execute(&self, input: Self::In) -> Result<Self::Out, Self::Err>;
+
+    /// Executes a compensation action for the operation.
+    ///
+    /// When using `infallible_transaction_with_strong_rollback_guarantees`, it is possible that there the compensation function
+    /// is called with no `result` provided.
+    /// It is the implementor's responsibility to decide if any action can be taken in such a case.
+    fn compensate(
+        &self,
+        input: Self::In,
+        result: Option<Result<Self::Out, Self::Err>>,
```

Comment:
I don't understand this. If the operation fails, then compensate should not be called, because there's nothing to compensate. So can't this be simplified to:

```suggestion
        result: Option<Self::Out>,
```

To eliminate the Option, could we just serialize `Out`, to make it:

```suggestion
        result: Self::Out,
```

Finally, how are you making this work without `Out` being Clone?

Reply:
Ok. I was assuming that we want the user to be able to perform compensation in case of an error too - so currently the first compensation action executed is the one for the failed operation, passing the failure cause to it.
I don't have a real-world example for it, it just felt right.

So if we say a failed operation never has to be compensated, then you are right in all points:
- we don't have to pass a `Result`
- and we can serialize `Out`s of the successful steps so there is no need for the `Option` either. In my version `None` would have been passed when we know an operation was started, but it never finished, to allow the user to try to compensate it somehow (check if it actually happened and do something conditionally, for example)

Should I make this simplification?

Reply:
The `Clone` constraint is there, just not on the type. Probably it should. (It's on the `execute` function)